### PR TITLE
Save app-metadata-file-output even if no change is present

### PR DESCRIPTION
**Pull request by praveenrewar**
#### What this PR does / why we need it:
Save app-metadata-file even if no change is present

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

##### Review Checklist:

- [x] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [x] Relevant tests are added or updated
- [x] Relevant docs in this repo added or updated
- [x] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [x] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kapp/cmd/app/deploy.go [hunk 1]
```diff
@@ -220,18 +222,7 @@ func (o *DeployOptions) Run() error {
...
-				if err != nil {
-					return
-				}
-				err = os.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)
-				if err != nil {
-					return
-				}
-			}
-		}()
+		defer o.writeAppMetadataToFile(app)
```
Comment (100mik) [inline pkg/kapp/cmd/app/deploy.go:225]: Is there any harm in doing this just once when we are sure we have the app (we can defer the function once, around line 116 I believe would work)?

I do not see why we *need* to write the file before `touch.Do` exits 🤔

**Review by 100mik — COMMENTED** [review]
Just a thought!

File: pkg/kapp/cmd/app/deploy.go (hunk 1 above)
Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go:225]: > Is there any harm in doing this just once when we are sure we have the app (we can defer the function once, around line 116 I believe would work)?

Good point!! I will check if that's doable.

> I do not see why we need to write the file before touch.Do exits 🤔

We need to save the metadata file even if there's no change to the app (we are returning early in that case and do not reach touch.Do)

Comment (100mik) [reply pkg/kapp/cmd/app/deploy.go:225]: > We need to save the metadata file even if there's no change to the app (we are returning early in that case and do not reach touch.Do)

Yep that makes sense. I was wondering if the deferring needs to be done inside the function passed to `touch.Do` for some reason instead of doing it once earlier, that is all. I could not think of anything right off the bat.

Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go:225]: So, we have the [AskForConformation](https://github.com/vmware-tanzu/carvel-kapp/blob/68322f5ba9dc153aac6b76d5ee97255883a7a46f/pkg/kapp/cmd/app/deploy.go#L190) on line 190 and I think that we shouldn't write to metadata before this, i.e. before the user gives the confirmation. The only special case to this would be when there are no changes or diff run flag is being used (which is where I have added the defer call). I am still thinking what would be the right place to have a single defer call, but do let me know if you have any thoughts on this.

Comment (100mik) [reply pkg/kapp/cmd/app/deploy.go:225]: Oh I see, so we only wanna write the file if
1. There are no changes
2. Changes get applied

Comment (100mik) [reply pkg/kapp/cmd/app/deploy.go:225]: The only change I see in that case is that when there are no changes, the call does not need to be deferred.

Another benefit of this would be that when we grab new GKs before confirmation, those do not get written unless the changes have been applied.

Resolving this and putting a comment above.

File: pkg/kapp/cmd/app/deploy.go [hunk 2]
```diff
@@ -178,6 +178,8 @@ func (o *DeployOptions) Run() error {
 	}
 
 	if o.DiffFlags.Run || hasNoChanges {
+		defer o.writeAppMetadataToFile(app)
```
Comment (100mik) [inline pkg/kapp/cmd/app/deploy.go]: Do we need to defer here?

Looks like we return after this no matter what. No room for erroring.

Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go]: Makes sense, updated.

File: test/e2e/app_change_test.go [hunk 3]
```diff
@@ -133,6 +133,8 @@ metadata:
 	assert.NoError(t, err)
 	secondDeploy, err := os.CreateTemp(os.TempDir(), "output2")
 	assert.NoError(t, err)
+	thirdDeploy, err := os.CreateTemp(os.TempDir(), "output3")
```
Comment (100mik) [inline test/e2e/app_change_test.go:136]: I think this entire test should go into a separate file, discussed this with @praveenrewar and that could go in as a separate PR to keep this one less noisy

@praveenrewar good first issue maybe? October is around the corner 😆

**Review by 100mik — APPROVED** [review]
LGTM!

File: test/e2e/app_change_test.go (hunk 3 above)
Comment (praveenrewar) [reply test/e2e/app_change_test.go:136]: Yeah, separate PR makes sense 👍🏻

**Comment by sethiyash** [comment]
Looks Good.

File: pkg/kapp/cmd/app/deploy.go [hunk 4]
```diff
@@ -472,6 +463,20 @@ func (o *DeployOptions) existingPodResources(existingResources []ctlres.Resource
 	return existingPods
 }
 
+func (o *DeployOptions) writeAppMetadataToFile(app ctlapp.App) error {
+	if o.DeployFlags.AppMetadataFile != "" {
+		meta, err := app.Meta()
+		if err != nil {
+			return err
+		}
+		err = os.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)
```
Comment (rohitagg2020) [inline pkg/kapp/cmd/app/deploy.go]: Change to `return os.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)`

Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go]: Done.

**Review by rohitagg2020 — APPROVED** [review]

**Comment by kumaritanushree** [comment]
LGTM

**Review by 100mik — APPROVED** [review]
LGTM 🚀

**Comment by 100mik** [comment]
Created #604 for some minor refactoring
